proto/pulumi/provider.proto

```diff
@@ -215,32 +220,105 @@ message GetSchemaResponse {
     string schema = 1; // the JSON-encoded schema.
 }
 
+// `ConfigureRequest` is the type of requests sent as part of a [](pulumirpc.ResourceProvider.Configure) call. Requests
+// include both provider-specific inputs (`variables` or `args`) and provider-agnostic ("protocol") configuration
+// (`acceptSecrets`, `acceptResources`, and so on).
 message ConfigureRequest {
-    // the input properties for the provider, with nested values JSON-encoded; please read from `args` instead.
+    // :::{warning}
+    // `variables` is deprecated; `args` should be used instead wherever possible.
+    // :::
+    //
+    // A map of input properties for the provider. Compound values, such as nested objects, should be JSON encoded so
+    // that they too can be passed as strings. For instance, the following configuration:
+    //
+    // ```
+    // {
+    //   "a": 42,
+    //   "b": {
+    //     "c": "hello",
+    //     "d": true
+    //   }
+    // }
+    // ```
+    //
+    // should be encoded as:
+    //
+    // ```
+    // {
+    //   "a": "42",
+    //   "b": "{\"c\":\"hello\",\"d\":true}"
+    // }
+    // ```
     map<string, string> variables = 1;
 
-    google.protobuf.Struct args = 2;   // the input properties for the provider
-    bool acceptSecrets  = 3;           // when true, operations should return secrets as strongly typed.
-    bool acceptResources = 4;          // when true, operations should return resources as strongly typed values to the provider.
-    bool sends_old_inputs = 5;         // when true, diff and update will be called with the old outputs and the old inputs.
-    bool sends_old_inputs_to_delete = 6; // when true, delete will be called with the old outputs and the old inputs.
+    // A map of input properties for the provider.
+    //
+    // :::{warning}
+    // `args` may include secrets. Because `ConfigureRequest` is sent before [](pulumirpc.ConfigureResponse) can specify
+    // whether or not the provider accepts secrets in general, providers *must* handle secrets if they appear in `args`.
+    // :::
+    google.protobuf.Struct args = 2;
+
+    // True if and only if the caller supports secrets. If true, operations should return strongly typed secrets if the
```

Comment:
As we know it is too late because secrets may have been sent in args? This is super quirky but might be worth a note that we recognize it is not ideal, that is a legacy situation.